Documentation/installation/kind.rst

```diff
@@ -48,36 +48,30 @@ Then, install Cilium release via Helm:
 
 .. note::
 
-   To enable Cilium's kube-proxy replacement (:ref:`kubeproxy-free`), kind nodes
-   need to have their own cgroup namespaces, and are different from the cgroup namespace
+   To enable Cilium's Socket LB (:ref:`kubeproxy-free`), Kind nodes
+   need to run in separate `cgroup namespaces
+   <https://man7.org/linux/man-pages/man7/cgroup_namespaces.7.html>`__,
+   and these namespaces need to be different from the cgroup namespace
    of the underlying host so that Cilium can attach BPF programs at the right
-   cgroup hierarchy. To verify this, run the following commands inside
-   respective kind nodes, and ensure that the cgroup values are different.
+   cgroup hierarchy. To verify this, run the following commands, and ensure
+   that the cgroup values are different:
 
    .. code-block:: shell-session
 
-      $ docker exec -it kind-control-plane bash
-      $ ls -al /proc/self/ns/cgroup
+      $ docker exec kind-control-plane ls -al /proc/self/ns/cgroup
       lrwxrwxrwx 1 root root 0 Jul 20 19:20 /proc/self/ns/cgroup -> 'cgroup:[4026532461]'
 
-      $ docker exec -it kind-worker bash
-      $ ls -al /proc/self/ns/cgroup
+      $ docker exec kind-worker ls -al /proc/self/ns/cgroup
       lrwxrwxrwx 1 root root 0 Jul 20 19:20 /proc/self/ns/cgroup -> 'cgroup:[4026532543]'
 
-   If you've deployed a kind cluster on an environment where kind nodes don't
-   have their own cgroup namespaces, then cgroup v2 needs to be enabled by
```

Comment:
The existing document does mention that the real requirement is separate cgroup namespaces in above few lines. But if it's not met, then we need to advise users that the way to have separate cgroup namespaces is to enable cgroup v2.

Reply:
Enabling Cgroup v2 does not guarantee that cgroup namespaces will be used for Kind node containers. In Docker this is controlled by `--default-cgroupns-mode` (https://docs.docker.com/engine/reference/commandline/dockerd/). For example, when running in cgroup v2, but with `--default-cgroupns-mode=host`, each container will end up in the same cgroup ns.

Reply:
That's why we've linked the PR 16259 in the same note, it has these details. We could update the existing step in the guide to enable cgroup v2 with something along these lines - "enable cgroup v2 whereby container runtimes enable create every pod in a separate namespace by default...". Without this step, the note looks incomplete.

Reply:
I think that PR is a bit too low-level for the Kind GSG audience.

Anyway, my aim is to clarify two requirements:
- cgroup namespaces (first paragraph)
- cgroup v1 interference (second)

> enable cgroup v2 whereby container runtimes enable create every pod in a separate namespace by default

Are you concerned that we don't mention cgroup v2 as a third requirement? If yes, the KPR GSG has it, and also cilium-agent will panic if cgroup v2 is disabled. Because of that we can keep the note concise.

Reply:
>If yes, the KPR GSG has it, and also cilium-agent will panic if cgroup v2 is disabled.

Can you share links? KPR has been working fine with cgroupv1 all along. Note that there is a difference between cgroupv2 controllers mode and cgroup2 fs. Cgroup2 mode (former) is relevant here in this guide.

Reply:
> Can you share links?

The note which starts with `Cilium will automatically mount cgroup v2 filesystem ...` from https://docs.cilium.io/en/v1.12/gettingstarted/kubeproxy-free/

> KPR has been working fine with cgroupv1 all along

Socket LB won't work if cgroup v2 is disabled.

Reply:
See my last comment - https://github.com/cilium/cilium/pull/20749#discussion_r936975525. 

Socket-lb has always worked with cgroup v1 mode/controllers? Cilium mounts **cgroup2 fs** which is an interface to kernel internals. But in this context, we are not talking about cgroup2 fs.

Reply:
> Socket-lb has always worked with cgroup v1 mode/controllers?

Socket LB always required cgroup v2. Cgroup v1 is important in Kind context because of its possible interference with v2 which got fixed in https://github.com/torvalds/linux/commit/8520e224f547cd070c7c8f97b1fc6d58cff7ccaa.

Missing the cgroup v2 support would result in failing to mount cgroup2 fs.

Reply:
I think we are going in circles. Linux kernel supports both cgroup v1 and cgroup v2, which is why cilium agent can mount cgroup2 fs on clusters running cgroup v1 controllers. Socket-lb has been working fine even before the recent K8s move to cgroup v2. 

Let me try to recap the issues and resolutions from PR 16259, some of which are relevant for this thread -

- In order for socket-lb to work in kind environments, the kind nodes need to have their own cgroup namespaces, separate from the host cgroup namespace. This ensures that cilium is able to attach cgroup BPF programs at the right hierarchy. This is evident from many reports where users are running with cgroupv1, and socket-lb doesn't work. 
  
- While cgroups v2 (controllers) isn't required to have separate cgroup namespaces, container runtimes switched to creating private cgroup namespaces for containers with cgroup v2. In case of docker, the relevant option is `default-cgroupns-mode` [1], for podman –cgroupns=mode [2], and so on.

I don't think kind exposes such control settings? It probably relies on the container runtime installed on the host. Regardless, we can't just assume cgroups v2, and not mention in the kind gsg. 

The current kind gsg already captures most of this, but we can add some more nuances to make the requirements more explicit. To avoid further back and forth on this PR, here is what I propose for updating the note - 

 ```
  If you've deployed a kind cluster on an environment where kind nodes don't
   have their own cgroup namespaces, then cgroup v2 needs to be enabled by
   setting the kernel ``systemd.unified_cgroup_hierarchy=1`` parameter. This 
   enables container runtimes such as docker [1], podman [2] to enable private
   cgroup namespaces for containers. 
   cgroup v1/v2 interference...
```


[2] https://docs.podman.io/en/v2.0.6/markdown/podman-create.1.html
–cgroupns=mode

Set the cgroup namespace mode for the container. host: use the host’s cgroup namespace inside the container. container:<NAME|ID>: join the namespace of the specified container. ns:: join the namespace at the specified path. private: create a new cgroup namespace.

If the host uses cgroups v1, the default is set to host. On cgroups v2 the default is private.

```

Reply:
Thanks for the detailed answer.

> I don't think kind exposes such control settings? It probably relies on the container runtime installed on the host.

Yes, it's controlled by container runtimes.

Do we agree on the following requirements for Socket LB to work on Kind?

1. cgroup v2 (generic)
2. cgroup_namespaces (Kind)
3. cgroup v1 net_cls and net_prio disabled if running on < 5.14 kernel (Kind)

The 1. one is generic, so it should be only in the KPR GSG. The 2. and 3. are Kind specific, thus they should be mentioned only in the Kind's GSG (I could extend the 2. mentioning that a container runtime should enable it).

Based on above, I think that the following is redundant in the Kind's GSG (not to mention that it's only for environments which run systemd):

> then cgroup v2 needs to be enabled by setting the kernel ``systemd.unified_cgroup_hierarchy=1`` parameter.

Reply:
Can you try out the steps below? 

- Pick an environment with cgroup v1 and kernel > 5.14 that has the cgroup v1/v2 interference fix
- Deploy a kind cluster, and check if socket-lb works reliably. 

This should settle whether/why we need the [note](https://github.com/cilium/cilium/pull/20749#discussion_r941577333) in the guide to enable cgroup v2.

Reply:
I don't have a machine at hand which would have both enabled. Also, I am not exactly following why checking the v1/v2 interference fix would resolve our previous discussion?